### ISSUE TITLE
[BUG] It's possible to break modmail's internal thread cache and simultanously open many threads

### ISSUE THREAD

**Issue by bast0006** [kyb3r/modmail#3022]
**Describe the bug**
If `confirm_thread_creation` is used, a user can break the thread creation flow and somehow cause modmail's DM response to freeze. This then results in modmail sending as many `confirm_thread_creation` prompts as messages were sent during the freeze, and if the user reacts to each one, they can open one thread per message. 

**Bot Info**
Bot version (check with `@modmail about`): 3.9.4
Host method (Heroku, self-host, etc): self-hosted

**To Reproduce**
1. Who can reproduce (ex. anyone, owners)? Anyone
2. Where can it be reproduced (ex. in thread channels, recipient DM's)? DM's
3. Done what to cause the error?
 .1. DM modmail, wait for the `confirm_thread_creation` message to come up. 
 .2. Block modmail
 .3. React to create the thread and close the thread before modmail can respond
 .4. Unblock modmail, and while it's frozen (~10-15s depending), send a few messages. The effect is worse if you wait a short time between each one
 .5. Observe _many_ `confirm_thread_creation` messages that all work. 
4. Any recently made changes to your bot? Some print statements, and requests hitting a channel category instead of the guild's text channels. 

**Error Logs**
https://paste.pythondiscord.com/luyevadeyu.txt

**Screenshots**
https://cdn.discordapp.com/attachments/620607373828030464/840607305586049034/Screen_Shot_test1.png
https://cdn.discordapp.com/attachments/620607373828030464/840607475187449866/Screen_Shot_test2.png
https://cdn.discordapp.com/attachments/620607373828030464/840607476554792990/Screen_Shot_test3.png

**Additional context**

**Comment by Jerrie-Aries** [comment]
> .3. React to create the thread and close the thread before modmail can respond

What do you mean by *close the thread before modmail can respond*?
Do you mean close the thread from thread channel (using command), or deny the thread creation (from DM)?

**Comment by bast0006** [comment]
Click both the check and X mark reactions before modmail is able to remove them

**Comment by Jerrie-Aries** [comment]
Thanks for the explanation.
I tried/tested it, and it seemed like I couldn't reproduce it.

**Comment by bast0006** [comment]
If you have an instance available for testing I'd love to reproduce it for you?

**Comment by Jerrie-Aries** [comment]
Sure, can you DM me on discord? `Jerrie#7234`.

**Comment by StephenDaDev** [comment]
I have been able to half reproduce this on the latest version, running on Heroku.

**Comment by StephenDaDev** [comment]
<img width="161" alt="Screen Shot 2021-05-10 at 10 51 46 AM" src="https://user-images.githubusercontent.com/48072084/117678792-b889d080-b17d-11eb-9b3c-263e3d4e5e1c.png">
I have been able to fully reproduce this issue on the latest version, running on Heroku.

**Comment by StephenDaDev** [comment]
2021-05-10 14:54:06 __main__[1331] - ERROR: Ignoring exception in on_message.
2021-05-10 14:54:06 __main__[1332] - ERROR: Unexpected exception:
Traceback (most recent call last):
  File "/app/.heroku/python/lib/python3.9/site-packages/discord/client.py", line 333, in _run_event
    await coro(*args, **kwargs)
  File "/app/bot.py", line 1007, in on_message
    await self.process_commands(message)
  File "/app/bot.py", line 1014, in process_commands
    return await self.process_dm_modmail(message)
  File "/app/bot.py", line 814, in process_dm_modmail
    thread = await self.threads.create(message.author, message=message)
  File "/app/core/thread.py", line 1207, in create
    await message.channel.send(
  File "/app/.heroku/python/lib/python3.9/site-packages/discord/abc.py", line 904, in send
    data = await state.http.send_message(channel.id, content, tts=tts, embed=embed,
  File "/app/.heroku/python/lib/python3.9/site-packages/discord/http.py", line 241, in request
    raise Forbidden(r, data)
discord.errors.Forbidden: 403 Forbidden (error code: 50007): Cannot send messages to this user
2021-05-10 14:54:17 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:31 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:31 core.thread[379] - ERROR: Thread already closed: 231595246213922828.
2021-05-10 14:54:31 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:31 core.thread[379] - ERROR: Thread already closed: 231595246213922828.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:35 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:36 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:36 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:37 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:37 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:37 core.thread[1074] - WARNING: Found existing thread for 231595246213922828 but the channel is invalid.
2021-05-10 14:54:38 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:40 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:43 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:44 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:45 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:47 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:49 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:50 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:51 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:52 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.
2021-05-10 14:54:53 __main__[1129] - WARNING: Failed to find linked message for reactions: Thread channel message not found.

Logs. Was able to reproduce a 2nd time.

**Comment by onerandomusername** [comment]
I can reproduce successfully multiple times in a row

![image](https://user-images.githubusercontent.com/71233171/117681229-e2dc8d80-b17f-11eb-89a9-aa662719dc3b.png)

**Comment by Jerrie-Aries** [comment]
> I can reproduce successfully multiple times in a row

Did the bot create multiple channels for you as well?

**Comment by StephenDaDev** [comment]
> On step three, don't even both reacting to open a thread, just react to close the thread. It works better, and is more reliable to recreate the issue.

Just did this and it spammed me with "Cancelled."

**Comment by onerandomusername** [comment]
> 
> 
> > On step three, don't even both reacting to open a thread, just react to close the thread. It works better, and is more reliable to recreate the issue.
> 
> Just did this and it spammed me with "Cancelled."

You still need to react with the checkbox for all later messages. The first one during the block didn't the checkbox tho.

**Comment by Jerrie-Aries** [comment]
Welp. I managed to reproduce this on my local machine.

![Untitled](https://user-images.githubusercontent.com/70805800/117688715-9e86d700-b1eb-11eb-91c4-9fcbcdccc3dc.png)

I had a patch ready for this issue a while back when I encountered similar issue using react to contact.

The issue is probably from here:
https://github.com/kyb3r/modmail/blob/3717df646644159c5a0e713aa83bfcdbf768dcdb/core/thread.py#L1120-L1128

Should also add `not thread.cancelled` as well in `if` statement to make sure there's no existing cancelled thread before returning *None*.
Returning None that makes the bot create a new thread.

Also need to handle the exception when sending "Cancelled" embed, this probably what causes the bot to hang without returning the Thread instance (which `cancelled` flag is set to True) to wherever it's called from.
